src/include/OSL/mask.h

```diff
@@ -0,0 +1,370 @@
+                                      Value16Type,
+                                      Value32or64Type>::type ValueType;
+#else
+    typedef Value32or64Type ValueType;
+#endif
+
+    static constexpr int width = WidthT;
+
+protected:
+    static constexpr int value_width = sizeof(ValueType) * 8;
+    static_assert(value_width >= WidthT, "unsupported WidthT");
+    static constexpr ValueType valid_bits
+        = static_cast<ValueType>(0xFFFFFFFFFFFFFFFF) >> (value_width - WidthT);
+
+public:
+    OSL_FORCEINLINE Mask() {}
+
+    explicit OSL_FORCEINLINE Mask(Lane lane) : m_value(1 << lane.value()) {}
+
+    explicit OSL_FORCEINLINE Mask(bool all_on_or_off)
+        : m_value((all_on_or_off) ? valid_bits : 0)
+    {
+    }
+
+    explicit constexpr OSL_FORCEINLINE Mask(std::false_type) : m_value(0) {}
+
+    explicit constexpr OSL_FORCEINLINE Mask(std::true_type)
+        : m_value(valid_bits)
+    {
+    }
+
+    explicit OSL_FORCEINLINE Mask(Value16Type value_)
+        : m_value(static_cast<ValueType>(value_))
+    {
+    }
+
+    explicit OSL_FORCEINLINE Mask(Value32Type value_)
+        : m_value(static_cast<ValueType>(value_))
+    {
+    }
+
+    explicit OSL_FORCEINLINE Mask(Value64Type value_)
+        : m_value(static_cast<ValueType>(value_))
+    {
+    }
+
+    explicit OSL_FORCEINLINE Mask(int value_)
+        : m_value(static_cast<ValueType>(value_))
+    {
+    }
+
+    OSL_FORCEINLINE Mask(const Mask& other) : m_value(other.m_value) {}
+
+    template<int OtherWidthT,
+             typename = pvt::enable_if_type<(OtherWidthT < WidthT)>>
+    explicit OSL_FORCEINLINE Mask(const Mask<OtherWidthT>& other)
+        : m_value(static_cast<ValueType>(other.value()))
+    {
+    }
+
+
+    OSL_FORCEINLINE ValueType value() const { return m_value; }
+
+    // count number of active bits
+    OSL_FORCEINLINE int count() const
+    {
+#if __INTEL_COMPILER
+        if (value_width <= 32)
+            return _mm_popcnt_u32(m_value);
+        else
+            return _mm_popcnt_u64(m_value);
+#else
+        // Compiler recognized idiom, can map to 1 instruction
+        ValueType m(m_value);
+        int count = 0;
+        for (count = 0; m != 0; ++count) {
+            m &= m - 1;
+        }
+        return count;
+#endif
+    }
+
+    // NOTE: undefined result if no bits are on
+    OSL_FORCEINLINE int first_on() const
+    {
+        // Simple naive implementation
+        // int lane=-1;
+        // while ( (m_value >> (++lane)) != 0 );
+        // return lane;
+
+        if (value_width <= 32) {
+#if __INTEL_COMPILER
+            return _bit_scan_forward(m_value);
+#else
+            return __builtin_ctz(m_value);
+#endif
+        } else if (value_width <= 64) {
+#if (defined(_WIN32) || defined(_WIN64))
+            unsigned long index;
+            _BitScanForward64(&index, m_value);
+            return static_cast<int>(index);
+#else  // Linux, OS X, gcc or clang
+            return __builtin_ctzll(m_value);
+#endif
```

Comment:
Let's change this to an `#if defined(__GNUC__) || defined(__clang__)`, then have a final `#else` clause with an appropriate `#error` inside, just in case anybody tries to compile this with some unforeseen compiler.

Reply:
sounds reasonable

Reply:
How's this:
```
#if (defined(_WIN32) || defined(_WIN64))
            unsigned long index;
            _BitScanForward64(&index, m_value);
            return static_cast<int>(index);
#elif defined(__GNUC__) || defined(__clang__)
            return __builtin_ctzll(m_value);
#else
#error unknown compiler, update Mask<int WidthT>::first_on() to support compiler
#endif
```

Reply:
looks good

Reply:
I tried to build OSL in Windows, from latest changes
this line broke the build
`OpenShadingLanguage\src\include\OSL/mask.h(166,20): error C3861: '__builtin_ctz': identifier not found `

Reply:
Oh yes, I see. Hang on, I will propose a fix...

Reply:
thanks @lgritz 
I can test it on Windows, any changes you apply, I am preparing Docs for building in windows, so can count on me on any OS based changes on the repo

Reply:
See #1291